presto-thrift-tpch-node/src/main/java/com/facebook/presto/thrift/node/ThriftServerTpch.java

```diff
@@ -0,0 +1,462 @@
+                indexInfo.getIndexableColumnNames());
+        checkState(indexedTableOptional.isPresent(), "index is not present");
+        TpchIndexedData.IndexedTable table = indexedTableOptional.get();
+        List<Type> keyTypes = types(indexInfo.getTableName(), table.getKeyColumns());
+        Page keysPage = convertToPage(keys, table.getKeyColumns(), keyTypes);
+        RecordSet keyRecordSet = new PageRecordSet(keyTypes, keysPage);
+        RecordSet allColumnsOutputRecordSet = table.lookupKeys(keyRecordSet);
+        List<Integer> outputRemap = computeRemap(table.getOutputColumns(), indexInfo.getOutputColumnNames());
+        RecordSet outputRecordSet = new MappedRecordSet(allColumnsOutputRecordSet, outputRemap);
+        return new ThriftSplitsOrRows(
+                null,
+                cursorToRowsBatch(
+                        outputRecordSet.cursor(),
+                        indexInfo.getOutputColumnNames(),
+                        estimateRecords(rowsMaxBytes, outputRecordSet.getColumnTypes()),
+                        continuationToken));
+    }
+
+    private ThriftRowsBatch getRowsInternal(byte[] splitId, long maxBytes, @Nullable byte[] continuationToken)
+    {
+        SplitInfo splitInfo = deserialize(splitId, SplitInfo.class);
+        List<String> columnNames = splitInfo.getColumnNames();
+        RecordCursor cursor = createCursor(splitInfo, splitInfo.getColumnNames());
+        return cursorToRowsBatch(
+                cursor,
+                columnNames,
+                estimateRecords(maxBytes, types(splitInfo.getTableName(), splitInfo.getColumnNames())),
+                continuationToken);
+    }
+
+    private static int estimateRecords(long maxBytes, List<Type> types)
+    {
+        if (types.isEmpty()) {
+            // no data will be returned, only row count
+            return Integer.MAX_VALUE;
+        }
+        int bytesPerRow = 0;
+        for (Type type : types) {
+            if (type instanceof FixedWidthType) {
+                bytesPerRow += ((FixedWidthType) type).getFixedSize();
+            }
+            else if (type instanceof VarcharType) {
+                VarcharType varchar = (VarcharType) type;
+                if (varchar.isUnbounded()) {
+                    // random estimate
+                    bytesPerRow += 32;
+                }
+                else {
+                    bytesPerRow += varchar.getLengthSafe();
+                }
+            }
+            else {
+                throw new IllegalArgumentException("Cannot compute estimates for an unknown type: " + type);
+            }
+        }
+        return toIntExact(maxBytes / bytesPerRow);
+    }
+
+    private static ThriftRowsBatch cursorToRowsBatch(
+            RecordCursor cursor,
+            List<String> columnNames,
+            int maxRowCount,
+            @Nullable byte[] continuationToken)
+    {
+        long skip = continuationToken != null ? Longs.fromByteArray(continuationToken) : 0;
```

Comment:
It depends on the continuation token type. The one for data uses long, the one for splits uses int.